File: .github/production_verify.mjs
import { chromium } from 'playwright';
import fs from 'fs';
const base='https://osaraclinics.com';
const browser=await chromium.launch({headless:true});
const errors=[]; const mobile=[];
const mobilePaths=['/','/dermatology','/ophthalmology','/doctors/dr-osama-alwreikat','/doctors/dr-sara-abu-touq','/vitiligo-jordan'];
for(const p of mobilePaths){
 const ctx=await browser.newContext({viewport:{width:390,height:844}}); const page=await ctx.newPage(); const ce=[];
 page.on('pageerror',e=>ce.push('pageerror: '+e)); page.on('console',m=>{if(m.type()==='error')ce.push('console: '+m.text())});
 const resp=await page.goto(base+p,{waitUntil:'networkidle',timeout:45000});
 const m=await page.evaluate(()=>({sw:document.documentElement.scrollWidth,bsw:document.body.scrollWidth,iw:innerWidth,broken:[...document.images].filter(i=>i.complete&&i.naturalWidth===0).map(i=>i.src),h1:document.querySelectorAll('h1').length}));
 mobile.push({path:p,status:resp?.status(),url:page.url(),...m,consoleErrors:ce});
 if(resp?.status()!==200||m.sw>390||m.bsw>390||m.broken.length||ce.length)errors.push('mobile '+p);
 await ctx.close();
}
const ctx=await browser.newContext({viewport:{width:1280,height:900}}); const page=await ctx.newPage(); const consoleErrors=[];
page.on('pageerror',e=>consoleErrors.push('pageerror: '+e)); page.on('console',m=>{if(m.type()==='error')consoleErrors.push('console: '+m.text())});
await page.addInitScript(()=>document.addEventListener('click',e=>{if(e.target.closest?.('a[href]'))e.preventDefault()},true));
await page.goto(base+'/',{waitUntil:'networkidle',timeout:45000});
const rendered=await page.evaluate(()=>{
 const body=document.body.innerText; let clinic=null;
 for(const s of document.querySelectorAll('script[type="application/ld+json"]')){try{const x=JSON.parse(s.textContent);if(x?.['@type']==='MedicalClinic'&&x?.['@id']==='https://osaraclinics.com/#clinic')clinic=x}catch{}}
 return {hours:clinic?.openingHoursSpecification||null,hoursText:document.querySelector('.hours-container')?.innerText||'',urgentAr:body.includes('تواصل للحالات العاجلة خارج ساعات العمل'),urgentEn:body.includes('For urgent dermatological or ophthalmological concerns outside regular clinic hours'),has24:/24\s*\/\s*7|24\s*hours/i.test(body)};
});
const expected=[{'@type':'OpeningHoursSpecification',dayOfWeek:['Saturday','Sunday'],opens:'16:30',closes:'20:30'},{'@type':'OpeningHoursSpecification',dayOfWeek:['Monday','Tuesday','Wednesday','Thursday'],opens:'08:00',closes:'13:30'},{'@type':'OpeningHoursSpecification',dayOfWeek:['Monday','Tuesday','Wednesday','Thursday'],opens:'16:30',closes:'20:30'}];
if(JSON.stringify(rendered.hours)!==JSON.stringify(expected))errors.push('rendered hours');
for(const x of ['Saturday - Sunday','Monday - Thursday','Friday','Closed','8:00 AM - 1:30 PM','4:30 PM - 8:30 PM'])if(!rendered.hoursText.includes(x))errors.push('visible hours '+x);
if(!rendered.urgentAr||!rendered.urgentEn||rendered.has24)errors.push('urgent wording');
async function testCTA(path,selector,name){await page.goto(base+path,{waitUntil:'networkidle',timeout:45000});await page.evaluate(()=>{window.gtag=undefined;window.dataLayer=[]});return await page.evaluate(({selector,name})=>{const a=document.querySelector(selector);if(!a)return {found:false,name};a.click();const dl=window.dataLayer||[];return {found:true,name,count:dl.filter(x=>x?.event===name).length,dataLayer:dl}}, {selector,name});}
const ctas=[];
ctas.push(await testCTA('/','a[href*="wa.me/"]:not([data-appointment]):not(.book-appointment)','whatsapp_click'));
ctas.push(await testCTA('/','a[href^="tel:"]','phone_click'));
ctas.push(await testCTA('/dermatology','a[href*="google.com/maps"]','directions_click'));
ctas.push(await testCTA('/','a[data-appointment],a.book-appointment','appointment_click'));
for(const c of ctas)if(!c.found||c.count!==1)errors.push('cta '+c.name);
if(consoleErrors.length)errors.push('desktop console');
const physician=[];
for(const p of ['/doctors/dr-osama-alwreikat','/doctors/dr-sara-abu-touq']){await page.goto(base+p,{waitUntil:'networkidle',timeout:45000});const x=await page.evaluate(()=>({imgs:[...document.images].map(i=>i.src),og:document.querySelector('meta[property="og:image"]')?.content||null,schema:[...document.querySelectorAll('script[type="application/ld+json"]')].some(s=>s.textContent.includes('"image"'))}));physician.push({path:p,...x});if(x.imgs.length||x.og||x.schema)errors.push('physician '+p)}
const report={mobile,rendered,ctas,physician,consoleErrors,errors};fs.writeFileSync('production-browser-report.json',JSON.stringify(report,null,2));console.log(JSON.stringify(report,null,2));await browser.close();if(errors.length)process.exit(1);

File: .github/production_verify.py
import json,re,sys
from urllib.parse import urljoin
import requests
from bs4 import BeautifulSoup
BASE='https://osaraclinics.com'
ROUTES=['/','/dermatology','/ophthalmology','/doctors/dr-osama-alwreikat','/doctors/dr-sara-abu-touq','/vitiligo-jordan','/botox-hyperhidrosis','/mole-removal','/school-health']
expected={p:(BASE+'/' if p=='/' else BASE+p) for p in ROUTES}
errors=[]; report={'routes':[]}
for p in ROUTES:
    r=requests.get(BASE+p,allow_redirects=False,timeout=30)
    soup=BeautifulSoup(r.text,'html.parser')
    c=soup.find('link',rel='canonical')
    canon=c.get('href') if c else None
    report['routes'].append({'path':p,'status':r.status_code,'canonical':canon})
    if r.status_code!=200: errors.append(f'{p} status {r.status_code}')
    if canon!=expected[p]: errors.append(f'{p} canonical {canon!r}')
robots=requests.get(BASE+'/robots.txt',timeout=30)
report['robots']={'status':robots.status_code,'text':robots.text}
if robots.status_code!=200 or 'Sitemap: https://osaraclinics.com/sitemap.xml' not in robots.text: errors.append('robots')
sm=requests.get(BASE+'/sitemap.xml',timeout=30)
report['sitemap_status']=sm.status_code
soup=BeautifulSoup(sm.text,'xml')
locs=[x.get_text(strip=True) for x in soup.find_all('loc')]
report['sitemap_urls']=locs
expected_locs=[expected[p] for p in ROUTES]
if sm.status_code!=200 or set(locs)!=set(expected_locs) or len(locs)!=9: errors.append('sitemap')
# homepage source schema
home=requests.get(BASE+'/',timeout=30).text
soup=BeautifulSoup(home,'html.parser'); clinic=None
for s in soup.find_all('script',type='application/ld+json'):
    try:x=json.loads(s.get_text())
    except Exception:continue
    if isinstance(x,dict) and x.get('@type')=='MedicalClinic' and x.get('@id')=='https://osaraclinics.com/#clinic': clinic=x
expected_hours=[
 {'@type':'OpeningHoursSpecification','dayOfWeek':['Saturday','Sunday'],'opens':'16:30','closes':'20:30'},
 {'@type':'OpeningHoursSpecification','dayOfWeek':['Monday','Tuesday','Wednesday','Thursday'],'opens':'08:00','closes':'13:30'},
 {'@type':'OpeningHoursSpecification','dayOfWeek':['Monday','Tuesday','Wednesday','Thursday'],'opens':'16:30','closes':'20:30'}]
report['source_hours']=clinic.get('openingHoursSpecification') if clinic else None
if not clinic or clinic.get('openingHoursSpecification')!=expected_hours: errors.append('source hours')
# physician source image checks
report['physicians']={}
for p in ['/doctors/dr-osama-alwreikat','/doctors/dr-sara-abu-touq']:
    t=requests.get(BASE+p,timeout=30).text; s=BeautifulSoup(t,'html.parser')
    og=bool(s.find('meta',attrs={'property':'og:image'})); imgs=[i.get('src') for i in s.find_all('img')]
    schema_img=False
    for sc in s.find_all('script',type='application/ld+json'):
        try:x=json.loads(sc.get_text())
        except:continue
        if '"image"' in json.dumps(x): schema_img=True
    report['physicians'][p]={'og_image':og,'img_tags':imgs,'schema_image':schema_img}
    if og or imgs or schema_img: errors.append(f'physician image {p}')
# internal links among target routes
broken=[]
for p in ROUTES:
    t=requests.get(BASE+p,timeout=30).text; s=BeautifulSoup(t,'html.parser')
    for a in s.find_all('a',href=True):
        h=a['href']
        if h.startswith('/') and not h.startswith('//'):
            path=h.split('#')[0].split('?')[0] or '/'
            rr=requests.get(BASE+path,allow_redirects=False,timeout=30)
            if rr.status_code>=400: broken.append((p,h,rr.status_code))
report['broken_internal_links']=broken
if broken: errors.append('internal links')
# www redirect
rw=requests.get('https://www.osaraclinics.com/',allow_redirects=True,timeout=30)
report['www']={'final_status':rw.status_code,'final_url':rw.url,'history':[(x.status_code,x.headers.get('location')) for x in rw.history]}
if rw.status_code!=200 or rw.url!='https://osaraclinics.com/' or len(rw.history)==0 or len(rw.history)>3: errors.append('www redirect')
report['errors']=errors
open('production-static-report.json','w').write(json.dumps(report,indent=2,ensure_ascii=False))
print(json.dumps(report,indent=2,ensure_ascii=False))
sys.exit(1 if errors else 0)
